pkg/storagedistribution/storagedistribution.go

```diff
@@ -0,0 +1,390 @@
+				return nil, &cloudops.ErrInvalidStoragePoolUpdateRequest{
+					Request: request,
+					Reason: fmt.Sprintf("requested resize operation type cannot be " +
+						"accomplished as no existing drives were provided"),
+				}
+			}
+
+			newMinDeltaCapacityPerDrive := newDeltaCapacity / request.CurrentDriveCount
+
+			logrus.Debugf("min delta capacity per drive needed: %d GiB num drives: %d",
+				newMinDeltaCapacityPerDrive, request.CurrentDriveCount)
+
+			// check in matrix if this drive can be resized to newMinCapacityPerDrive
+			newMinCapacityPerDrive := request.CurrentDriveSize + newMinDeltaCapacityPerDrive
+			logrus.Debugf("need to resize drive to atleast: %d GiB", newMinCapacityPerDrive)
+			instStorage, err := instanceStorageForRow(row, newMinCapacityPerDrive, 0)
+			if err != nil {
+				if err == cloudops.ErrStorageDistributionCandidateNotFound {
+					continue ROW_LOOP
+				}
+
+				return nil, err
+			}
+			instStorage.DriveCount = request.CurrentDriveCount
+			resp.InstanceStorage = append(resp.InstanceStorage, instStorage)
+
+			return resp, nil
+		default:
+			// TODO try resize first. If doesn't satisfy, try add
+			return nil, &cloudops.ErrNotSupported{
+				Operation: fmt.Sprintf("GetStorageUpdateConfig with operation type: %d",
+					request.ResizeOperationType),
+			}
+		}
+	}
+
+	return nil, cloudops.ErrStorageDistributionCandidateNotFound
+}
+
+func getStorageDistributionCandidate(
+	decisionMatrix *cloudops.StorageDecisionMatrix,
+	request *cloudops.StorageSpec,
+	requestedInstancesPerZone uint64,
+	zoneCount uint64,
+) (*cloudops.StoragePoolSpec, uint64, error) {
+	logDistributionRequest(request, requestedInstancesPerZone, zoneCount)
+
+	if zoneCount <= 0 {
+		return nil, 0, cloudops.ErrNumOfZonesCannotBeZero
+	}
+
+	filteredRows := filterDecisionMatrix(decisionMatrix, request.IOPS, request.DriveType)
+	// Calculate min capacity per zone
+	minCapacityPerZone := request.MinCapacity / uint64(zoneCount)
+
+	for _, row := range filteredRows {
+		var (
+			capacityPerNode  uint64
+			instancesPerZone uint64
+		)
+		for instancesPerZone = requestedInstancesPerZone; instancesPerZone > 0; instancesPerZone-- {
+			capacityPerNode = minCapacityPerZone / uint64(instancesPerZone)
+			printCandidates("Candidate", []cloudops.StorageDecisionMatrixRow{row}, instancesPerZone, capacityPerNode)
+
+			instStorage, err := instanceStorageForRow(row, capacityPerNode, 0)
+			if err != nil {
+				if err == cloudops.ErrStorageDistributionCandidateNotFound {
+					printCandidates("Candidate failed to satisfy requirements.",
+						[]cloudops.StorageDecisionMatrixRow{row}, instancesPerZone, capacityPerNode)
+					continue
+				}
+
+				logrus.Errorf("failed to get storage spec for instance: %v", err)
+				return nil, 0, cloudops.ErrStorageDistributionCandidateNotFound
+			}
+
+			// additional check so that we are not overprovisioning
+			for instancesPerZone > 1 && instStorage.DriveCapacityGiB*instancesPerZone > minCapacityPerZone {
```

Comment:
Do we need this? 
Can we just check if the cluster capacity is less than maxCapacityPerZone otherwise the outer for loop already reduces the instancesPerZone and recalculates.

Also you should consider DriveCount as well `instStorage.DriveCapacityGiB*instancesPerZone*instStorage.DriveCount`